freud/locality.pyx

```diff
@@ -198,7 +198,8 @@ cdef class NeighborQueryResult:
 
         npoint = dereference(iterator).next()
         while npoint != ITERATOR_TERMINATOR:
-            yield (npoint.query_point_idx, npoint.point_idx, npoint.distance)
+            yield (npoint.getQueryPointIdx(), npoint.getPointIdx(),
+                   npoint.getDistance())
```

Comment:
One more point for discussion: When iterating over a `NeighborQueryResult` in python: In addition to returning the two indices, should we now return the vector, the vector/distance, the distance, or return the vector/weight/distance? 

My personal vote is for at least returning the vector in some form, although I'm not very opinionated on the other data we return with it.

Reply:
I could see us doing the full tuple where users can pattern match to grab what they want.

Reply:
In case you're curious: when we made the `NeighborList` iterable, we only returned the point indices because we thought those would be the most commonly requested. It's already possible to do something like:
```python
for bond_idx, (query_idx, point_idx) in enumerate(nlist):
    weight = nlist.weights[bond_idx]
    print(query_idx, point_idx, weight)
```
(Untested snippet, based on my recollection of the API).

Reply:
@b-butler I'm not super familiar with the developer side of pattern matching, what would that look like implementation-wise?

Reply:
just a `yield query_idx, point_idx, vector, distance, weight`

Reply:
For sake of completeness, we can just have it return all 5 values in case a user wants to use them. What @bdice says about iterating over the `NeighborList` is true, but I don't think the same holds for iterating over a `NeighborQueryResult`.

Reply:
I think it would be best if iterating over either object type gave the same kinds of information. `NeighborQueryResult` is kind of like a lazy/generated `NeighborList`.

Reply:
Since the iterating is done lazily in `NeighborQueryResult`, we cannot index into the `NeighborList` to get all 5 pieces of information like we can with the `NeighborList`. To make iterating over the `NeighborQueryResult` the most useful, we would need to return all 5 pieces of information. If we decide we want to keep it symmetric between `NeighborList` and `NeighborQueryResult`, that would mean that iterating over a `NeighborList` should return all 5 pieces of information also.

In general, I don't think users find the `NeighborQueryResult` to be a very useful object in python (I've had people in lab ask me why it even exists in the first place), so I don't have a huge opinion on whether they should match or not. It also takes very little time to generate the full `NeighborList` from the `NeighborQueryResult`, so after thinking it over again, I'm okay leaving everything as is for now.

Reply:
I originally added the `NeighborQueryResult` in #241 as a way for users to prototype new neighbor-based analysis methods in Python. It's a more Pythonic approach than building and iterating over a `NeighborList` since it functions more naturally as an iterator. In practice I don't think anyone has really been doing new methods development so the use case probably hasn't really come up. For that reason, I agree that most users wouldn't find it useful, any more than most users find iterating directly over a `NeighborList` useful.

As far as making iteration over a `NeighborList` and a `NeighborQueryResult` identical, I agree that would be optimal but there are meaningful differences here. A `NeighborList` can encapsulate the other data into members/properties, whereas the result object cannot, so there's good reason for it to behave differently. The alternative would be to implement `NeighborList.__iter__` and explicitly have it return all 5 properties instead of having it fall back to `__getitem__`. That raises questions about whether it's odd to have `next(iter(nlist))` return something different than `nlist[0]`, though, so there's no perfect solution, but that may be the best compromise.

Reply:
I would like to provide a little further argumentation for the Python layer treating (query) point indices as “the iterable thing” and not including weight/distance/vector in the iterable. The indices are the most fundamental part of the neighbor list concept in my eyes, and all else is auxiliary. I think it’s sufficient for users to zip-iterate over multiple arrays of indices/weights/etc as needed. The easy path `for i, j in nlist` should keep it minimal and only yield indices.

Reply:
I'm inclined to agree that we shouldn't change how NeighborList iteration behaves. However, that doesn't address the case where no NeighborList is constructed. The implication is that the only way to iterate over the extra arrays (weights/distances/vectors) is to construct a `NeighborList`, i.e. there's no way to do this without materialization. I don't really like that, but that's the tradeoff for keeping the iteration consistent with `NeighborList`.

One other option (not necessarily advocating for it, just throwing out ideas) is taking a page from numpy's book and adding `return_*` arguments (or maybe a single `return_opts={key: val}` parameter) like `numpy.unique`.